service/history/historyEngine.go

```diff
@@ -573,7 +573,7 @@ Update_History_Loop:
 		var failCause workflow.DecisionTaskFailedCause
 		var err error
 		completedID := *completedEvent.EventId
-		hasUnhandledEvents := ((completedID - startedID) > 1)
+		hasUnhandledEvents := (msBuilder.GetNextEventID() - completedID) > 1
```

Comment:
Two things:
1) We should first append all the decisions as part of decision task completed before flushing buffer to history events.
2) Then this check could just be msBuilder.HasBufferedEvents.

Reply:
will update